kubessh/sftp.py

```diff
@@ -0,0 +1,164 @@
+        """,
+        config=True
+    )
+
+    # where the SFTP binary goes in the user pod
+    sftp_binary_upload_path = Unicode(
+        '.local/sbin/kubessh-sftp-server',
+        allow_none=False,
+        help="""
+        Path inside the user pod where the sftp-server binary will be uploaded.
+
+        Must be full path including the executable name, and can be absolute or
+        relative to the working dir. Must be writable by the user the pod runs
+        as.
+        """,
+        config=True
+    )
+
+    def __init__(self, server, reader, writer):
+        from kubessh.app import app
+        super(LoggingConfigurable, self).__init__(parent=app)
+        super().__init__(server, reader, writer)
+
+    def _run_setup_command(self, *kubectl_command):
+        self.logger.info(f'executing: {kubectl_command}')
+        cmd = subprocess.Popen(kubectl_command)
+        return cmd.wait()
+
+    def _init_delegate(self):
+        self.namespace = self.parent.default_namespace
+        self.pod_name = UserPod(
+            parent=self.parent, namespace=self.namespace,
+            username=self.username).pod_name
+        # ensure sftp-server binary is copied into the user pod
+        self._run_setup_command(
+            'kubectl',
+            '--namespace', self.namespace,
+            'exec',
+            '-c', 'shell',
+            self.pod_name,
+            '--', 'mkdir', '-p',
+            os.path.dirname(self.sftp_binary_upload_path))
+        self._run_setup_command(
+            'kubectl', 'cp',
+            self.sftp_binary_source_path,
+            f'{self.namespace}/{self.pod_name}:{self.sftp_binary_upload_path}',
+            '-c', 'shell')
+        self._run_setup_command(
+            'kubectl',
+            '--namespace', self.namespace,
+            'exec',
+            '-c', 'shell',
+            self.pod_name,
+            '--', 'chmod', '+x', self.sftp_binary_upload_path)
+
+        # start sftp-server binary and redirect stdin/stdout
+        kubectl_command = [
+            'kubectl',
+            '--namespace', self.namespace,
+            'exec',
+            '-c', 'shell',
+            '--stdin',
+            self.pod_name,
+            '--', self.sftp_binary_upload_path
+        ]
+        self.logger.info(f'starting delegate sftp-server process: {kubectl_command}')
+
+        self._proc = subprocess.Popen(
+            kubectl_command,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=sys.stderr
+        )
+        self._proc_stdin = self._proc.stdin
+        self._proc_stdout = self._proc.stdout
+
+
+
+## XXX: as of asyncssh v2.2.1, there appears to be no proper way to override
+## the default SFTPServerHandler implementation. The below monkey-patching
+## works for now, but should be replaced once there is a better way to do this.
+
+import asyncssh.stream
+import asyncssh.sftp
+
+def run_override(sftp_server, reader, writer):
```

Comment:
Do you think we can turn this into a function we explicitly call somewhere, instead of automatically running on import? I don't have much experience with monkey patching, however - so if you think this is the best way to do this, I'm ok with that.

Reply:
I think it's fine to have this in a function, probably better to make it explicit